server/src/main/java/org/elasticsearch/search/query/QueryPhaseCollectorManager.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.search.query;
+
+import org.apache.lucene.search.Collector;
+import org.apache.lucene.search.CollectorManager;
+import org.apache.lucene.search.Weight;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+
+class QueryPhaseCollectorManager<TD, AG> implements CollectorManager<QueryPhaseCollector, QueryPhaseCollectorManager.Result<TD, AG>> {
```

Comment:
the types are slightly controversial right now. They ease testing for now. Especially the aggs now we are not going to leverage in practice outside of tests. I am thinking that the fact that they help testing is a good reason to keep them, but I am open to other ideas.

Reply:
Do we really need to return that Result object? I prefer tp keep doing the same thing as we were doing before and register results on the query result object instead of trying to return the values in such a way. 

Sorry again, but I struggle to see the benefits of generics here.

Reply:
I am hoping that you'll see the benefits when I am done integrating it in the query phase, which this PR does not do.

Bear with me, a couple more PRs are needed to complete this work and the types are so far used to simplify testing and retrieve the results from the two collector managers. That is why I left myself a question asking for feedback. I can look into removing the type for aggs, as it is not strictly needed, although not having it will complicate a bit retrieving the aggs result in the test.

The long term plan is to remove the indirections / supplier in TopDocsCollectorManagerFactory: based on the collector chosen, we build total hits, and top docs differently. All that can be integrated into the query phase collector manager, so that the search method returns the actual thing that gets set to the query result. Right now, we set the top docs to the query result but it is hard to follow where the top docs come from, as it is done through the suppliers. I think that this will make things better compared to how they currently are.

That is my reasoning for why we may not need generics for aggs, but we will use them at our advantage for aggs. Plus I don't see how not to deal with generics  entirely when we rely on Lucene design that does have a certain contract / constraints which revolves around the reduce method.

Reply:
> the same thing as we were doing before

To clarify, what we had in previous prototypes required saving the results of the reduction in the collector manager and adding methods to retrieve it, while generics allow us to retrieve the results directly from the search method. This may not work with aggs, as I understand they may not rely on calling reduce but for top docs collection, it is beneficial.